devicetypes/stelpro/stelpro-ki-zigbee-thermostat.src/stelpro-ki-zigbee-thermostat.groovy

```diff
@@ -0,0 +1,667 @@
+				map.unit = ""
+			}
+			else if (descMap.value == "8000") {	//0x8000
+				map.name = "temperatureAlarm"
+				map.value = "cleared"
+				map.unit = ""
+			}
+			
+			else if (descMap.value > "8000") {
+				map.value = -(Math.round(2*(655.36 - map.value))/2)
+			}
+		}
+		else if (descMap.cluster == "0201" && descMap.attrId == "0012") {
+			log.debug "HEATING SETPOINT"
+			map.name = "heatingSetpoint"
+			map.value = getTemperature(descMap.value)
+			map.data = [heatingSetpointRange: heatingSetpointRange]
+			if (descMap.value == "8000") {		//0x8000
+				map.name = "temperatureAlarm"
+				map.value = "cleared"
+				map.data = []
+			}
+		}
+		else if (descMap.cluster == "0201" && descMap.attrId == "001c") {
+			if (descMap.value.size() == 8) {
+				log.debug "MODE"
+				map.name = "thermostatMode"
+				map.value = modeMap[descMap.value]
+				map.data = [supportedThermostatModes: supportedThermostatModes]
+			}
+			else if (descMap.value.size() == 10) {
+				log.debug "MODE & SETPOINT MODE"
+				def twoModesAttributes = descMap.value[0..-9]
+				map.name = "thermostatMode"
+				map.value = modeMap[twoModesAttributes]
+				map.data = [supportedThermostatModes: supportedThermostatModes]
+			}
+		}
+		else if (descMap.cluster == "0201" && descMap.attrId == "401c") {
+			log.debug "SETPOINT MODE"
+			log.debug "descMap.value $descMap.value"
+			map.name = "thermostatMode"
+			map.value = modeMap[descMap.value]
+			map.data = [supportedThermostatModes: supportedThermostatModes]
+		}
+		else if (descMap.cluster == "0201" && descMap.attrId == "0008") {
+			log.debug "HEAT DEMAND"
+			map.name = "thermostatOperatingState"
+			if (descMap.value < "10") {
+				map.value = "idle"
+			}
+			else {
+				map.value = "heating"
+			}
+
+			if (settings.heatdetails == "No") {
+				map.displayed = false
+			}
+		}
+	}
+
+	def result = null
+	if (map) {
+		result = createEvent(map)
+	}
+	log.debug "Parse returned $map"
+	return result
+}
+
+def updateWeather() {
+	log.debug "updating weather"
+	def weather
+	// If there is a zipcode defined, weather forecast will be sent. Otherwise, no weather forecast.
+	if (settings.zipcode) {
+		log.debug "ZipCode: ${settings.zipcode}"
+		weather = getWeatherFeature( "conditions", settings.zipcode )
+
+		// Check if the variable is populated, otherwise return.
+		if (!weather) {
+			log.debug( "Something went wrong, no data found." )
+			return false
+		}
+		
+		def tempToSend
+		
+		if(getTemperatureScale() == "C" ) {
+			tempToSend = weather.current_observation.temp_c
+			log.debug( "Outdoor Temperature: ${tempToSend} C" )
+		}
+		else {
+			tempToSend = weather.current_observation.temp_f
+			log.debug( "Outdoor Temperature: ${tempToSend} F" )
+		}	   
+		sendEvent( name: 'outsideTemp', value: tempToSend )
+		quickSetOutTemp(tempToSend)
+	}
+}
+
+def scheduledUpdateWeather() {
+	def actions = updateWeather()
+
+	if (actions) {
+		sendHubCommand(actions)
+	}
+}
+
+def getModeMap() { [
+	"00":"off",
+	"04":"heat",
+	"05":"eco"
```

Comment:
The spec says this is "Emergency Heating", but I'm guessing this is just a manufacturer quirk

Reply:
It is indeed :)